toplevel/topdirs.ml

```diff
@@ -122,18 +130,24 @@ let _ = add_directive "cd" (Directive_string dir_cd)
       doc = "Change the current working directory.";
     }
 
-let dir_load ppf name = ignore (Topeval.load_file false ppf name)
 
-let _ = add_directive "load" (Directive_string (dir_load std_out))
+let with_error f x = f (error_fmt ()) x
+let action_on_suberror b = if not b then fail ()
+
+let dir_load ppf name =
+  action_on_suberror (Topeval.load_file false ppf name)
+
+let _ = add_directive "load" (Directive_string (with_error dir_load))
     {
       section = section_run;
       doc = "Load in memory a bytecode object, produced by ocamlc.";
     }
 
-let dir_load_rec ppf name = ignore (Topeval.load_file true ppf name)
+let dir_load_rec ppf name =
+  action_on_suberror (Topeval.load_file true ppf name)
 
 let _ = add_directive "load_rec"
-    (Directive_string (dir_load_rec std_out))
+    (Directive_string (with_error dir_load_rec))
```

Comment:
Note: there is something subtly wrong with the fact that directives take a single formatter, on which they only print errors. This is the property that you use in this refactoring, but it does not hold for all directives, which is why your PR needs special treatment of `#help` and `#show`. (What is the expected behavior of `#show_val fubar;;` in non-interactive mode?)

I think that there would be various ways to improve on this:
1. We could pass two different formatters to our directives, an output formatter and an error formatter.
2. We could make the code simpler, instead of more complex, by changing the `directive_fun` types to add `oppf:Format.formatter -> eppf:Format.formatter ->`, and let Toploop pass the formatters.
3. Alternatively, the directives could be in charge of deciding which formatter to use, and call `error_formatter ()` directly. (But this makes it harder to redirect directive output to a completely different place.)

I'm not so fond of (3). (2) sounds nice, but it may have unpleasant backward-compatibility implications for utop and down. It may be that the best compromise is to remain "subtly wrong" and not change the interface at all, as this PR is doing, but this is worth a discussion.

Reply:
I tend to agree that morally directives should take both an error and output formatters as argument, and the current set of directives accidentally only use one of them.

> (What is the expected behavior of #show_val fubar;; in non-interactive mode?)

I would expect the result of `show` to be print on `stdout` as part of the script result.

Reply:
In this example, `fubar` is missing from the environment, I wondered if it should be counted as a directive failure or not. (Currently it is not, I guess?)

Reply:
Currently, it is not. And keeping it that way seems somewhat sensible (if certainly a bit arbitrary): sucesses or failures of `#show` do not alter the state of the script. Or looking from another perspective, a failure to find an identifier is still a valid answer to the user's query "show me some information about `fubar`".